rfc/plugins.md

```diff
@@ -0,0 +1,206 @@
+proceeds immediately.
+
+If a promise from any of the plugins is rejected then the deployment
+pipeline will stop and ember-cli-deploy will exit. Returned promises that are
+rejected are treated as unrecoverable errors.
+
+### Configuration
+
+By convention, plugin configuration should be kept in `config/deploy.js` and scoped by
+the plugin's name. e.g. for an `ember-cli-deploy-example` plugin, the configuration would
+go in:
+
+```javascript
+// config/deploy.js
+module.exports = {
+  production: {
+    "example": {
+      bucket: "my-app-production",
+      awsAccessKeyId: process.env.AWS_ACCESS_KEY_ID,
+      awsSecretAccessKey: process.env.AWS_SECRET_ACCESS_KEY
+    }
+  }
+};
+```
+
+### Hooks
+
+These hooks (part of a typical deployment process) are available for plugins to implement:
+
+```
+willDeploy: --> runs before anything happens. good opportunity for plugins to validate
+                configuration or other preconditions
+
+           /-- willBuild    confirm environment
+          /
+build --------> builds app assets, documentation, etc.
+          \
+           \-- didBuild     manipulate index.html, validate assets
+
+           /-- willUpload   confirm remote servers(S3, Redis, Azure, etc.)
+          /
+upload -------> puts the assets somewhere(S3, Redis, Azure, Rackspace, etc.)
+          |     Note: a plugin that implements upload of the HTML file and
+          |           wants to support version activation should set
+          |           `currentVersion` on the `deployment` object to the ID
+          |           of the newly deployed version.
+          \
+           \-- didUpload    notify APIs (slack, pusher, etc.), warm cache
+
+           /-- willActivate   create backup of assets, notify APIs, uninstall earlier versions
+          /
+activate -------> make a new version live (clear cache, swap Redis values, etc.)
+          \
+           \-- didActivate    notify APIs, warm cache
+
+  Note: when hooks in the activate series of hooks are called, the plugin can assume the
+        presence of a `currentVersion` property on the deployment object, that is set to
+        the ID of the version to be activated.
+
+didDeploy: --> runs at the end of a full deployment operation.
+```
+
+In addition, there are a few more specialized hooks that plugins may implement:
+
+```
+discoverVersions: --> should return a promise resolving to an array of version objects. Each
+                      version object _must_ have an `id` property. Each version _may_ have one
+                      or more of the following properties:
+
+                      `timestamp`:   (Date) when the version was created
+                      `revision`:    (String) reference of version in SCM
+                      `creator`:     (String) email address of developer who deployed the version
+                      `description`: (String) summary of the version
+
+```
+
+### Advanced Plugin Configuration
```

Comment:
I really like this section for the advanced plugin configuration stuff! I think it really addresses the use cases that have come up lately

:+1: